File: Java/Data Structures/Trees/BinaryTrees/Binary-Tree-Preorder-Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<Integer>values;
    
    public Solution(){
        this.values = new ArrayList<>();
    }
    public List<Integer> preorderTraversal(TreeNode root) {
        if(root==null){
            return this.values; 
        }
        if(root!=null){
            this.values.add(root.val);
        } 
        this.preorderTraversal(root.left);
        this.preorderTraversal(root.right);
        return this.values;
    }
}

File: Leetcode Learning/Data Structures/Trees/Binary Tree/Traverse a Tree/Binary-Tree-Pre-order-Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.values = []
    def preorderTraversal(self, root: TreeNode) -> List[int]:        
        if not root:
            return self.values
        if root:
            self.values.append(root.val)
        self.preorderTraversal(root.left)    
        self.preorderTraversal(root.right)    
        return self.values        
